resourcename/scanner.go

```diff
@@ -0,0 +1,65 @@
+package resourcename
+
+import (
+	"strings"
+)
+
+// Scanner scans a resource name.
+type Scanner struct {
+	name                     string
+	start, end               int
+	serviceStart, serviceEnd int
+	full                     bool
+}
+
+// Init initializes the scanner.
+func (s *Scanner) Init(name string) {
+	s.name = name
+	s.start, s.end = 0, 0
+	s.full = false
+}
+
+// Scan to the next segment.
+func (s *Scanner) Scan() bool {
+	switch s.end {
+	case len(s.name):
+		return false
+	case 0:
+		// Special case for full resource names.
+		if strings.HasPrefix(s.name, "//") {
+			s.full = true
+			s.start, s.end = 2, 2
+			nextSlash := strings.IndexByte(s.name[s.start:], '/')
+			if nextSlash == -1 {
```

Comment:
when would this code path be triggered?

Reply:
When the resource name has a service but nothing else, so this test:

~~~go
		{
			name:        "full without segments",
			input:       "//library.googleapis.com",
			full:        true,
			serviceName: "library.googleapis.com",
			segments:    []Segment{},
		},
~~~

Reply:
Gotcha! Had already pressed send before seeing the tests 🙉